File: src/algorithm/StudentAssignment.cpp
//
// Created by vkolobara on 12/25/17.
//

#include "StudentAssignment.h"

void StudentAssignment::assign(Parser parser) {
    auto busCapacity = parser.getBusCapacity();
    auto maxWalk = parser.getMaxWalk();
    const auto &busStops = parser.getBusStops();
    const auto &students = parser.getStudents();
    const auto &school = parser.getSchool();
    const auto &stopDistances = parser.getStopToSchoolDistance();
    const auto &studentDistances = parser.getStudentToStopDistance();

    auto alpha = 20;

    auto cmp = [&](int left, int right) {
        return stopDistances[left] < stopDistances[right];
    };

    busStopCount = std::vector<int>(busStops.size(), 0);

    for (int i=0; i<students.size(); i++) {
        std::priority_queue<int, std::vector<int>, decltype(cmp)> queue(cmp);

        for (int j=0; j<busStops.size(); j++) {
            if (maxWalk >= studentDistances[i][j]) {
                queue.push(j);
            }
        }

        if (queue.size() == 1) {
            assignment[i] = queue.top();
        } else {
            std::vector<int> possibilities;

            for (int j=0; j<alpha && !queue.empty(); j++) {
                possibilities.push_back(queue.top());
                queue.pop();
            }
            std::uniform_int_distribution<> distr(0, possibilities.size()-1);
            int ind = distr(Random::eng);
            int sel = possibilities[ind];

            while(true) {
                if (busStopCount.at(sel) < busCapacity) {
                    assignment[i] = sel;
                    break;
                } else {
                    possibilities.erase(possibilities.begin()+ind);
                    ind = std::max(ind-1, 0);
                    sel = possibilities[ind];
                }
            }
        }
        busStopCount[assignment[i]]++;


    }
}

StudentAssignment::StudentAssignment() {

}

const std::map<int, int> &StudentAssignment::getAssignment() const {
    return assignment;
}

const std::vector<int> &StudentAssignment::getBusStopCount() const {
    return busStopCount;
}

File: src/hr/vinko/hmo/v1/algorithm/StudentAssignment.java
package hr.vinko.hmo.v1.algorithm;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Random;
import java.util.stream.IntStream;

import hr.vinko.hmo.parser.Coordinate;
import hr.vinko.hmo.parser.Parser;

public class StudentAssignment {
	private Map<Integer, Integer> assignment;
	private int[] busStopCount;

	private static Random rand = new Random();

	public StudentAssignment(int nStops) {
		assignment = new HashMap<>();
		busStopCount = new int[nStops];
	}

	public static StudentAssignment assign(Parser parser, boolean greedy) {

		int busCapacity = parser.getBusCapacity();
		double maxWalk = parser.getMaxWalk();
		List<Coordinate> busStops = parser.getBusStops();
		List<Coordinate> students = parser.getStudents();
		// Coordinate school = parser.getSchool();
		double[] stopDistances = parser.getStopToSchoolDistance();
		double[][] studentDistances = parser.getStudentToStopDistance();

		StudentAssignment assignment = new StudentAssignment(busStops.size());

		int alpha;
		if (greedy) {
			alpha = Math.max(1, busStops.size()/5);
		} else {
			alpha = busStops.size();
		}
		
		int studList[] = IntStream.range(0, students.size()).toArray();

		for (int i = 0; i < studList.length; i++) {
			int randomPosition = rand.nextInt(studList.length);
			int temp = studList[i];
			studList[i] = studList[randomPosition];
			studList[randomPosition] = temp;
		}

		for (int i = 0; i < studList.length; i++) {
			int index = studList[i];
			List<Integer> queue = new ArrayList<>();

			for (int j = 0; j < busStops.size(); j++) {
				if (maxWalk >= studentDistances[index][j])
					queue.add(j);
			}

			if (queue.size() == 1)
				assignment.assignment.put(index, queue.get(0));
			else {
				queue.sort(new Comparator<Integer>() {
					@Override
					public int compare(Integer o1, Integer o2) {
						return Double.compare(stopDistances[o1], stopDistances[o2]);
					}
				});
				int ind = 0;

				ind = rand.nextInt(Math.min(alpha, queue.size()));

				while (true) {
					int sel = queue.get(ind);

					if (assignment.busStopCount[sel] < busCapacity) {
						assignment.assignment.put(index, sel);
						break;
					} else {
						queue.remove(ind);
						ind = ind - 1 > 0 ? ind - 1 : 0;
					}
				}

			}

			assignment.busStopCount[assignment.assignment.get(index)]++;
		}

		return assignment;

	}

	public Map<Integer, Integer> getAssignment() {
		return assignment;
	}

	public int[] getBusStopCount() {
		return busStopCount;
	}

	@Override
	public String toString() {
		StringBuilder sb = new StringBuilder();

		for (Entry<Integer, Integer> entry : assignment.entrySet()) {
			sb.append((entry.getKey() + 1) + " " + (entry.getValue() + 1) + "\n");
		}

		return sb.toString();

	}

}
